商品購入機能

takanori-hagiwara opened the pull request:
# WHAT
furima app 商品購入機能の実装

#  WHY
GYAZO
正しく情報を記入すると、商品の購入ができる動画
https://gyazo.com/6b63bfe7e327de356b9f78b69b63de0b
入力に問題のある状態では商品の購入はできず、エラーメッセージが出力される動画https://gyazo.com/0fd0bb71d917eae82de81e07c9975ccf
ログアウト状態のユーザーが、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画 
https://gyazo.com/1bbfaff538bc1b493df720c43689cf6f
ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
https://gyazo.com/e6c677f3a00ecfc82b456000d0c92def
ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品一覧、詳細機能実装時に未実装であった場合）
https://gyazo.com/db8e0883a591ff078181132e2796a7ae
ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/8156df88981128d01c503272a462dfcb
ログイン状態の出品者以外のユーザーでも、売却済みの商品に対しては「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/f898e9f06223e173cf9ff8f6570a6bfc
ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/073df0100f29dc7fe788c44f6db87f98
テスト結果の画像
https://gyazo.com/5f3c611bb806eda0902cf6aff143609c

追記
- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/068ad6a4bd846925ff89c79bf1950723

takanori-hagiwara committed:
購入画面の表示

takanori-hagiwara committed:
商品購入データの保存

takanori-hagiwara committed:
クレジットカード決済の実装

takanori-hagiwara committed:
テストコードの実装

chisesakai commented:
レビュー依頼ありがとうございます！キャプチャ拝見させていただきました。
いよいよ購入機能ですね。引き続き一緒に頑張っていきましょう！

**- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画**
恐れ入りますが、上記キャプチャが確認できませんでしたので、改めてキャプチャの撮影・提出お願いいたします。

添付頂いた後、再度レビュー依頼のご提出をお願い致します！

takanori-hagiwara committed:
商品購入機能 修正1

- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/068ad6a4bd846925ff89c79bf1950723

kuu0123 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,43 @@
+  before_action :contributor_confirmation
+
+  def index
+    @item = Item.find(params[:item_id])
+    @order_address = OrderAddress.new
+  end
+
+  def create
+    @item = Item.find(params[:item_id])
```
```ruby
	    @item = Item.find(params[:item_id])
```
現在、index,create,ontributor_confirmationメソッドに`@item`が記載されているようです。
こちらはset_itemなどのメソッドを作成し、before_actionで呼び出すようにしましょう。
DRY（Don't Repeat Yourself、繰り返しを避けること）を意識するとコードの可読性が上がります。

kuu0123 commented on app/views/items/index.html.erb:
```diff
@@ -131,9 +131,11 @@
         <div class='item-img-content'>
           <%= image_tag item.image, class: "item-img" %>
           <%# 商品が売れていればsold outを表示しましょう %>
```
「〜〜しましょう」というコメントアウトは実装案内用となり、実装後は不要な記述となります。
上記に合致する箇所は削除しておきましょう。

kuu0123 commented on app/views/items/show.html.erb:
```diff
@@ -23,17 +23,16 @@
+   <% end %>
   <% end %>
 
     <%# //ログインしているユーザーと出品しているユーザーが、同一人物の場合と同一人物ではない場合で、処理を分けましょう %>
```
「〜〜しましょう」というコメントアウトは実装案内用となり、実装後は不要な記述となります。
上記に合致する箇所は削除しておきましょう。

kuu0123 commented on spec/factories/order_addresses.rb:
```diff
@@ -0,0 +1,10 @@
+  factory :order_address do
+    postal_code { '123-4567' }
+    city { '福岡市' }
+    street_address { 1 }
```
```ruby
	    street_address { 1 }
```
こちらは文字列のカラムを用意しましょう。
クオテーションがない場合、数値（integer）として取り扱うテストになってしまいます。

kuu0123 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      sleep(1)
+    end
+
+    context '商品がが購入できる場合' do
```
建物番号は空でも登録できることを確認する正常系テストを実装しましょう！

kuu0123 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,68 @@
+        expect(@order_address.errors.full_messages).to include('Postal code is invalid')
+      end
+
+      it 'phoneが空では購入できない' do
```
電話番号については以下のテストも実装しましょう。
```
・英数混合では登録できないこと
```

kuu0123 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,68 @@
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Token can't be blank")
+      end
+    end
```
モデルにuser_idとitem_idのバリデーションを設定しているため、テストにも以下の項目を実装しましょう！
```
・user_idが空では購入できないこと
・item_idが空では購入できないこと
```

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

キャプチャ動画の追加分、確認いたしました。
7点コメントいたしましたので、ご確認お願いします！

takanori-hagiwara committed:
商品機能 修正2

Gyazo
テストコード結果
https://gyazo.com/ffa11cecad1dff2612d640c206683833

kuu0123 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def contributor_confirmation
+    @item = Item.find(params[:item_id])
```
```ruby
	    @item = Item.find(params[:item_id])
```
こちらの`@item`についても、`set_item`の記載を使用しましょう。

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
追加で1点コメントいたしましたので、ご確認お願いします！

takanori-hagiwara committed:
商品購入機能 修正3

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
こちらLGTMとさせていただきます！実装お疲れさまでした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
必須機能のレビューはこれで完了となります！
本当にお疲れ様でした！
